docs/guides/field-extensions.md

```diff
@@ -0,0 +1,172 @@
+
+    def apply(self, field: StrawberryField) -> None:
+        field.directives.append(Cached(time=self.caching_time))
+
+    def resolve(
+        self, next_: Callable[..., Any], source: Any, info: Info, **kwargs
+    ) -> Any:
+        current_time = time.time()
+        if self.last_cached + self.caching_time > current_time:
+            return self.cached_result
+        self.cached_result = next_(source, info, **kwargs)
+        return self.cached_result
+```
+
+```python+schema
+@strawberry.type
+class Client:
+
+    @strawberry.field(extensions=[CachingExtensions(caching_time=200)])
+    def analyzed_hours(self, info) -> int:
+        return do_expensive_computation()
+---
+type Client {
+    analyzedHours: Int! @Cached(time=200)
+}
+```
+
+## Combining multiple field extensions
+
+When chaining multiple field extensions, the outermost extension is called first,
+and then it calls the next extension until it reaches the resolver.
+The return value of each extension is passed as an argument to the next extension.
+This allows for creating a chain of field extensions that each perform a specific
+transformation on the data being passed through them.
+
+<Tip>
+
+**Order matters**: the outermost extension will be executed first, while the innermost
+extension will be applied to the field first. This enables cases like
+adding relay pagination in front of an extension that modifies the field's type.
+
+For a of `extensions` of length `n`, `extensions[0]` is the innermost extension,
+while `extensions[n-1]` is the outermost.
+
+</Tip>
+
+## Async Extensions and Resolvers
+
+Field Extensions support async execution using the `resolve_async` method.
+A field extension can either support `sync`, `async`, or both. The appropriate resolve
+function will be automatically chosen based on the type of resolver and other
+extensions.
+
+Since sync-only extensions cannot await the result of an async resolver,
+they are not compatible with async resolvers or extensions.
+
+The other way around is possible: you can add an async-only extension to a sync
+resolver, or wrap sync-only extensions with it.
+This is enabled by an automatic use of the `SyncToAsyncExtension`. Note that after
+adding an async-only extension, you cannot wrap it with a sync-only extension anymore.
+
+<Tip>
+
+To optimize the performance of your resolvers, it's recommended that you implement both
+the `resolve` and `resolve_async` methods when using an extension on both sync and async
+fields.
+While the `SyncToAsyncExtension` is convenient, it may add unnecessary
+overhead to your sync resolvers, leading to slightly decreased performance.
+
+</Tip>
+
+```python
+import strawberry
+from strawberry.extensions import FieldExtension
+
+
+class UpperCaseExtension(FieldExtension):
+    def resolve(self, next: Callable[..., Any], source: Any, info: Info, **kwargs):
+        result = next(source, info, **kwargs)
+        return str(result).upper()
+
+    async def resolve_async(
```

Comment:
we might want to make a decision on the names here, in our schema we have schema.execute and schema.execute_sync, could do we do the same here to keep consistency?
/cc @BryceBeagle @jkimbo

Reply:
To clarify why I chose to "flip" the naming: having `resolve` with async as the default case may cause more use of async-only extension on synchronous fields, because the naming implies that `resolve_sync` is the special case that is only needed when doing sync execution (even though the performance disadvantages of converting every resolver to async are stated in the docs - but we know that not everyone reads them). When the sync case is the default case, it might encourage more to implement the sync extension, too. 
That being said, I'm all for consistency, so I'm fine with adjusting to what you decide for ☺️